File: Tree/Easy/111-Minimum-Depth-of-Binary-Tree/solution_bfs.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int MinDepth(TreeNode root) {
        // bfs + level traversal
        // tc:O(n); sc:O(n)
        if(root == null) {
            return 0;
        }
        Queue<TreeNode> queue = new Queue<TreeNode>();
        int level = 0;
        queue.Enqueue(root);
        while(queue.Count > 0) {
            int levelSize = queue.Count;  // get count of each level
            level++;
            for(int i = 0; i < levelSize; i++) {
                TreeNode node = queue.Dequeue();
                if(node.left == null && node.right == null) {
                    return level;
                }
                if(node.left != null) {
                    queue.Enqueue(node.left);
                }
                if(node.right != null) {
                    queue.Enqueue(node.right);
                }
            }
        }
        return level;
    }
}

File: BFS/Medium/Lnt618-Search-Graph-Nodes/solution_bfs.cpp
/**
 * Definition for Undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */


class Solution {
public:
    /*
     * @param graph: a list of Undirected graph node
     * @param values: a hash mapping, <UndirectedGraphNode, (int)value>
     * @param node: an Undirected graph node
     * @param target: An integer
     * @return: a node
     */
     UndirectedGraphNode* searchNode(vector<UndirectedGraphNode*>& graph,
                                    map<UndirectedGraphNode*, int>& values,
                                    UndirectedGraphNode* node,
                                    int target) {
        // bfs level order traversal
        // tc:O(n*m); sc:O(n*m)
        if(graph.size() == 0) {
            return nullptr;
        }
        unordered_set<UndirectedGraphNode*> visited;
        queue<UndirectedGraphNode*> q;
        q.push(node);
        while(!q.empty()) {
            int levelSize = q.size();
            UndirectedGraphNode* nd = q.front();
            q.pop();
            if(values[nd] == target) {
                return nd;
            }
            visited.insert(nd);
            for(auto ngb : nd->neighbors) {
                if(visited.find(ngb) == visited.end()) {
                    q.push(ngb);
                }
            }
        }
        return nullptr;
        
    }
};